### PR TITLE
Refactor shell code in testsuite

### PR DESCRIPTION
**Pull request by praveenkumar**
- Create shell instance only for the command execute
- Remove channel dependency to get the exitCode
- Use go capability to get exitcode for a process instead running a custom command
- Refactor shell struct configuration

### CODE REVIEW

**Comment by cfergeau** [comment]
>     * Create shell instance only for the command execute
> 
>     * Remove channel dependency to get the exitCode
> 
>     * Use go capability to get exitcode for a process instead running a custom command
> 
>     * Refactor shell struct configuration

Reading the PR description, I was expecting several commits to cover these to make it easier to understand which hunk is related to which of these 4 changes?

**Comment by guillaumerose** [comment]
Why do we need a shell after all ? Can't we just run the command ?

**Comment by praveenkumar** [comment]
> Why do we need a shell after all ? Can't we just run the command ?

@guillaumerose When you say run the command does that mean directly `exec.Command()` or some other way?

**Comment by guillaumerose** [comment]
Yes, `exec.Command("crc", "setup")`

**Comment by praveenkumar** [comment]
> Yes, exec.Command("crc", "setup")

In that case we need to change all our integration test definition where we use something like `executing "sudo podman build -t hello:test -f Dockerfile" succeeds`, in the background we take whatever part of `" <> "` and pass it to shell so we don't need to parse the name and command args.

**Comment by guillaumerose** [comment]
We can do it automatically maybe? Parse the string and use the first element as cmd but...

In CRC there is this kind of test:
```
        When executing "POD=$(oc get pod -o jsonpath="{.items[0].metadata.name}")" succeeds
        And executing "echo $POD" succeeds
        And executing "oc delete pod $POD --now" succeeds
```

Impossible to do without a shell, so let's keep the shell until we have ginkgo :)

**Comment by praveenkumar** [comment]
> I was expecting several commits to cover these to make it easier to understand which hunk is related to which of these 4 changes?

@cfergeau These are not exactly different subset of change but a refactor around shell which I am not able to do in different commit messages without breaking a previous commit. What I can do is add another function just to do what `Execute` now doing => delete old suff and rename that function to `Execute`. I am not able to break this tightly coupled commit.

File: testsuite/shell.go [hunk 1]
```diff
@@ -128,103 +88,62 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
...
 	}
 
-	exitCode := <-shell.exitCodeChannel
-	shell.excbuf.WriteString(exitCode)
-
-	return err
+	go func() {
+		defer inPipe.Close()
+		io.WriteString(inPipe, command+"\n")
+	}()
```
Comment (cfergeau) [inline testsuite/shell.go]: `cmd.Stdin =  strings.NewReader(command+"\n")` might work here?

File: testsuite/shell.go [hunk 2]
```diff
@@ -17,53 +17,34 @@ limitations under the License.
...
-	errbuf   bytes.Buffer
-	excbuf   bytes.Buffer
-
-	outPipe io.ReadCloser
-	errPipe io.ReadCloser
-	inPipe  io.WriteCloser
-
-	outScanner *bufio.Scanner
-	errScanner *bufio.Scanner
+	checkExitCodeCmd int
```
Comment (cfergeau) [inline testsuite/shell.go]: This should be renamed.

File: testsuite/shell.go [hunk 3]
```diff
@@ -17,53 +17,34 @@ limitations under the License.
...
-	zshExitCodeCheck        = "echo %v$?"
-	cmdExitCodeCheck        = "echo %v%%errorlevel%%"
-	powershellExitCodeCheck = "echo %v$lastexitcode"
+	"github.com/cucumber/messages-go/v10"
 )
 
 var (
-	shell ShellInstance
+	exitError *exec.ExitError
+	pathError *os.PathError
```
Comment (cfergeau) [inline testsuite/shell.go]: These do not need to be global variables.

**Comment by cfergeau** [comment]
> > I was expecting several commits to cover these to make it easier to understand which hunk is related to which of these 4 changes?
> 
> @cfergeau These are not exactly different subset of change but a refactor around shell which I am not able to do in different commit messages without breaking a previous commit. What I can do is add another function just to do what `Execute` now doing => delete old suff and rename that function to `Execute`. I am not able to break this tightly coupled commit.

I wonder if the exit code stuff could be separated out... However, it might be the commit log that is missing details on the extent of the changes :) And why they are being done. Regarding the changes, it seems this switches from having some kind of 'shell server' which can be fed commands to execute, and returns the result, to direct invocation of the command, at the cost of starting the shell once per command?
I think you did that because of a bug which you were experiencing in CI, but the commit log says nothing about it, while it's not far from being the most important thing to have for future reference (why is this being done?).

**Comment by praveenkumar** [comment]
> I wonder if the exit code stuff could be separated out.

@cfergeau Updated the commit logs.  If I do that then I have to keep the old methods to get the exitcode which also doesn't work in case of `podman build` command.

File: features/testsuite.feature [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
 
+  @linux @darwin @windows
   Scenario: Contains
      When executing "go help" succeeds
      Then stdout should contain
      """
      Go is a tool for managing Go source code.
      """
 
+  @linux @darwin
```
Comment (cfergeau) [inline features/testsuite.feature:12]: Why? The commit log does not say anything about this.

File: appveyor.yml [hunk 5]
```diff
@@ -4,4 +4,4 @@ environment:
   GOPATH: c:\gopath
 stack: go 1.14
 test_script:
-  - go test
+  - go test --godog.tags=windows
```
Comment (cfergeau) [inline appveyor.yml:7]: Should this be squashed with the commit which added the @linux,@windows, ... tags?

File: features/testsuite.feature [hunk 6]
```diff
@@ -64,7 +64,7 @@ Quentin check whether their testsuite works properly.
       When deleting directory "newdir" succeeds
       Then directory "newdir" should not exist
 
-   # Config 
+   # Config
```
Comment (cfergeau) [inline features/testsuite.feature:105]: These changes are unrelated.

File: testsuite/shell.go [hunk 7]
```diff
@@ -128,103 +85,57 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
...
 		}
 	}
-
 	return
 }
 
-func StartHostShellInstance(shellName string) error {
-	return shell.Start(shellName)
+func SetUpHostShellInstance(shellName string) error {
+	return shell.SetUp(shellName)
```
Comment (cfergeau) [inline testsuite/shell.go]: Maybe `Setup` rather than `SetUp`?

File: testsuite/shell.go [hunk 8]
```diff
@@ -128,103 +85,57 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
...
-
-	go shell.ScanPipe(shell.outScanner, &shell.outbuf, "stdout")
-	go shell.ScanPipe(shell.errScanner, &shell.errbuf, "stderr")
-
-	err = shell.instance.Start()
-	if err != nil {
-		return err
-	}
 
 	fmt.Printf("The %v instance has been started and will be used for testing.\n", shell.name)
```
Comment (cfergeau) [inline testsuite/shell.go]: This message can probably be dropped. It at least needs to be reworded, as nothing is started anymore.

File: testsuite/shell.go [hunk 9]
```diff
@@ -128,103 +85,57 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
...
-
-	return err
+	return nil
 }
 
 func ExecuteCommand(command string) error {
-	if shell.instance == nil {
+	if shell == nil {
 		return errors.New("shell instance is not started")
 	}
```
Comment (cfergeau) [inline testsuite/shell.go:112]: I don't think `shell` can be `nil`. `shell.name` could be empty though. And the `started` reference is obsolete.

File: testsuite/shell.go [hunk 10]
```diff
@@ -128,103 +85,57 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
...
-		return err
-	}
-
-	_, err = shell.inPipe.Write([]byte(shell.checkExitCodeCmd + "\n"))
-	if err != nil {
-		return err
+	cmd := exec.Command(shell.name, shell.startArgument...)
+	cmd.Stderr = &shell.errbuf
+	cmd.Stdout = &shell.outbuf
+	cmd.Stdin = strings.NewReader(command + "\n")
```
Comment (cfergeau) [inline testsuite/shell.go:126]: The log says
> We still create the input Pipe and pass the command in a goroutine but just after that
executing this stdinPipe closed 

I think this is no longer  correct?

File: testsuite/shell.go [hunk 11]
```diff
@@ -128,103 +85,57 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
...
+
+	if err := cmd.Run(); err != nil {
+		switch {
+		case errors.As(err, &exitError):
+			shell.exitCode = exitError.ExitCode()
+		case errors.As(err, &pathError):
+			return fmt.Errorf("os.PathError: %w", err)
+		default:
+			return fmt.Errorf("something went worng with %s command: %w", command, err)
```
Comment (cfergeau) [inline testsuite/shell.go]: 'wrong'

File: appveyor.yml (hunk 5 above)
Comment (praveenkumar) [reply appveyor.yml:7]: This is specific to CI so I have as separate commit.
